packages/core/src/middleware/koa-console-redirect-proxy.test.ts

```diff
@@ -0,0 +1,54 @@
+import { ossConsolePath } from '@logto/schemas';
+import type { Next } from 'koa';
+
+import { MockQueries } from '#src/test-utils/tenant.js';
+import { createContextWithRouteParameters } from '#src/utils/test-utils.js';
+
+import koaConsoleRedirectProxy from './koa-console-redirect-proxy.js';
+
+const { jest } = import.meta;
+
+const hasActiveUsers = jest.fn();
+const queries = new MockQueries({
+  users: { hasActiveUsers },
+});
+const next: Next = jest.fn(); // Mock the behaviour of next
+
+// tests for the koaConsoleRedirectProxy function.
+describe('koaConsoleRedirectProxy()', () => {
+  afterEach(() => {
+    // Performed after each test
+    jest.clearAllMocks();
+  });
+
+  it("should not call next() if ctx.path is '/' or '/welcome' and should redirect with 'ossConsolePath/welcome if hasUser is false'", async () => {
```

Comment:
can make the description of each unit test case more clear:
call next() function is not a result, but means redirect is not applied. So if you mention that “the page has been redirected”, then the next() function will never be called in this case.
for example, you can say -- should redirect to /welcome if there is no user in the database and the page is at the root path.